pkg/syncer/cnsoperator/controller/cnsregistervolume/cnsregistervolume_controller.go

```diff
@@ -326,17 +330,51 @@ func (r *ReconcileCnsRegisterVolume) Reconcile(ctx context.Context,
 
 	if syncer.IsPodVMOnStretchSupervisorFSSEnabled {
 		if workloadDomainIsolationEnabled || len(clusterComputeResourceMoIds) > 1 {
-			azClustersMap := topologyMgr.GetAZClustersMap(ctx)
-			isAccessible := isDatastoreAccessibleToAZClusters(ctx, vc, azClustersMap, volume.DatastoreUrl)
-			if !isAccessible {
-				log.Errorf("Volume: %s present on datastore: %s is not accessible to any of the AZ clusters: %v",
-					volumeID, volume.DatastoreUrl, azClustersMap)
-				setInstanceError(ctx, r, instance, "Volume in the spec is not accessible to any of the AZ clusters")
-				_, err = common.DeleteVolumeUtil(ctx, r.volumeManager, volumeID, false)
+			if isMultipleClustersPerVsphereZoneEnabled {
+				// Get zones assigned to the namespace
+				zoneMaps := commonco.ContainerOrchestratorUtility.GetZonesForNamespace(request.Namespace)
+
+				// Convert map keys to slice
+				zones := make([]string, 0, len(zoneMaps))
+				for zone := range zoneMaps {
+					zones = append(zones, zone)
+				}
+				// Get active clusters in the requested zones
+				activeClusters, err := commonco.ContainerOrchestratorUtility.
+					GetActiveClustersForNamespaceInRequestedZones(ctx, request.Namespace, zones)
 				if err != nil {
-					log.Errorf("Failed to untag CNS volume: %s with error: %+v", volumeID, err)
+					msg := fmt.Sprintf("Failed to get active clusters for CnsRegisterVolume request. error: %+v", err)
+					log.Error(msg)
+					setInstanceError(ctx, r, instance, msg)
+					return reconcile.Result{RequeueAfter: timeout}, nil
+				}
+				// Check if volume is accessible in any of the active clusters
+				if !isDatastoreAccessibleToAZClusters(ctx, vc,
+					map[string][]string{"dummy-zone-key": activeClusters}, volume.DatastoreUrl) {
+					log.Errorf("Volume: %s present on datastore: %s is not accessible to any of the active "+
+						"cluster on the namespace: %v", volumeID, volume.DatastoreUrl, activeClusters)
+					setInstanceError(ctx, r, instance, "Volume in the spec is not accessible to any of "+
+						"the active cluster on the namespace")
+
+					// Attempt volume cleanup
+					if _, err = common.DeleteVolumeUtil(ctx, r.volumeManager, volumeID, false); err != nil {
+						log.Errorf("Failed to untag CNS volume: %s with error: %+v", volumeID, err)
+					}
+					return reconcile.Result{RequeueAfter: timeout}, nil
```

Comment:
Let's make this a permanent failure and not requeue.

Reply:
done, if delete volume fails adding back to Requeue for reconcilation.
other wise returning  return reconcile.Result{}, nil